docs/configuration.md
# Configuration

Schism Tracker saves its configuration in a plain-text,
[INI-style](http://en.wikipedia.org/wiki/INI_file) file named `config`. Under
normal circumstances it should theoretically not be necessary to deal with this
file directly. However, there are some options that are not configurable from
within Schism Tracker for one reason or another. In these cases, any plain text
editor should suffice.

The location of this file is dependent on the OS:

- **Windows**
  - `%APPDATA%\Schism Tracker`
- **Mac OS X**
  - `~/Library/Application Support/Schism Tracker`
- **AmigaOS 4**
  - `PROGDIR:`
- **Linux/Unix**
  - `$HOME/.schism/`
- **Wii**
  - Same directory as `boot.elf` (e.g. `sd:/apps/schismtracker`)

Aside from `config`, you may also add a `fonts` subdirectory for custom font
files. By default, `font.cfg` is automatically loaded on startup, and other
`*.itf` files are listed by the built-in font editor (Shift-F12). See [the
links page](https://github.com/schismtracker/schismtracker/wiki/Links) for some
resources on getting fonts for Schism Tracker.

## Potentially useful "hidden" config options

To enable any of these, find the `[section]` in the config file, look for the
`key=`, and change the value. If the key doesn't exist, simply add it.

#### Video

    [Video]
    lazy_redraw=1
    width=640
    height=400
    want_fixed=0
    gl_bilinear=1

`lazy_redraw` slows down the framerate when the program isn't focused. This
used to be kind of useful when the GUI rendering sucked, and maybe it still is
if you're stuck with a painfully slow video card and software rendering, and
you also want to have a huge window that isn't active.

`width` and `height` are the initial dimensions to use for the window, and the
dimensions to return to when toggling fullscreen off.

If you are having problems with fullscreen aspect ratio, you can try setting
`want_fixed` to 1. This will attempt to "correct" the fullscreen resolution
used by Schism.

If `gl_bilinear` is zero, OpenGL will not use bilinear filtering when scaling
the screen. This has no effect when running at the native resolution of
640x400.

#### Backups

    [General]
    make_backups=1
    numbered_backups=1

When overwriting a `filename.it`, copy the existing file to `filename.it~`.
With numbered_backups, write to `filename.it.1~`, `filename.it.2~`, etc.

#### Key repeat

    [General]
    key_repeat_delay=125
    key_repeat_rate=25

Alter the key repeat. "Delay" is how long before keys begin to repeat, "rate"
is how long between repeated keystrokes. (Both are in milliseconds.) Above are
[Storlek](https://github.com/schismtracker/schismtracker/wiki/Storlek)'s
settings, which are very fast but convenient for speed tracking.

The *default* repeat delay and rate come from your operating system, so you
only need to set this if you like having a different rate for Schism Tracker
than you do for the rest of your system.

#### Alternate font

    [General]
    font=notch.itf

Load some other font besides `font.cfg` at startup. This option doesn't really
have much of a point, because the file listed is limited to those within the
`fonts` directory, and it's just easier to open the font editor, browse fonts,
and save to `font.cfg`.

#### DJ mode

    [General]
    stop_on_load=0

If zero, loading a song when another one is playing will start playing the new
song after it is loaded.

#### Startup behavior

    [General]
    bypass_splash=0

By default Schism Tracker starts up to the about page before continuing.
When `bypass_splash` is set to 1, Schism Tracker will skip this and startup directly to the load module page.

#### File browser

    [Directories]
    module_pattern=*.it\073 *.xm\073 *.s3m\073 *.mtm\073 *.669\073 *.mod

Changes what files are presented in the load/save module lists. Use * for all
files. For annoying compatibility reasons, semicolons are rewritten as `\x3b`
or `\073` when saving.

This was formerly named `filename_pattern`; Schism Tracker ignores the old
value and comments it out when saving to work around bugs in older versions.

    sort_with=strcasecmp

Alter the sort order. Possible values are `strcmp` (case-sensitive),
`strcasecmp` (case-insensitive), and `strverscmp` (case-sensitive, but handles
numbers smartly e.g. `5.it` will be listed above `10.it`).

#### Keyjazz

    [Pattern Editor]
    keyjazz_noteoff=1
    keyjazz_write_noteoff=0
    keyjazz_repeat=0

If `keyjazz_noteoff` is 1, letting go of a key in the pattern editor will cause
a note-off. If using this, you might also want to consider setting
`keyjazz_repeat` to 0 in order to avoid inserting multiple notes when holding
down keys.

If `keyjazz_write_noteoff` is 1, letting go of a key in the pattern editor will
also write a note off *if* playback tracing (Ctrl+F) is enabled.

#### Pattern editor behavior tweaks

    [Pattern Editor]
    mask_copy_search_mode=1
    invert_home_end=1

When `mask_copy_search_mode` is set to 1, pressing Enter on a row with no
instrument number will search backward in the channel for an instrument and
switch to that one.

`invert_home_end` changes the order of the Home and End keys to make the cursor
move to the first or last row within the channel before moving to the first or
last channel. FT2 users might want to enable this.

#### Key modifiers

    [General]
    meta_is_ctrl=1
    altgr_is_alt=1

These alter how modifier keys are interpreted. Mac OS X users in particular
might appreciate `meta_is_ctrl`, which allows using the Command/Apple key as
the Ctrl modifier within Schism Tracker. `altgr_is_alt` works similarly.

#### Audio output

    [Audio]
    buffer_size=256
    driver=dsp:/dev/dsp1

These settings define the audio buffer size, and which audio device Schism
Tracker uses. (The other settings in the `[Audio]` section are configurable
from Shift-F1.) `buffer_size` should be a power of two and defines the number
of samples in the mixing buffer. Smaller values result in less audio latency
but could cause buffer underruns and skipping.

`driver` is parsed identically to the `--audio-driver` switch on the command
line. If you're using Alsa on Linux and want to use you can set
`driver=alsa:dmix` to get Schism Tracker to play with other programs. (However,
Alsa completely ignores the latency with dmix so it might cause massive delays
between pressing a note and hearing it, which is why Schism Tracker requests a
"real" device by default.)

    [Diskwriter]
    rate=96000
    bits=16
    channels=2

This defines the sample format used by the disk writer – for exporting to
.wav/.aiff *and* internal pattern-to-sample rendering.

## Hook functions

Schism Tracker can run custom scripts on startup, exit, and upon completion of
the disk writer. These are stored in the configuration directory, and are named
`startup-hook`, `exit-hook`, and `diskwriter-hook` respectively. (On Windows,
append `.bat` to the filenames.) Hooks are useful for making various
adjustments to the system – adjusting the system volume, remapping the
keyboard, etc. The disk writer hook can be used to do additional
post-processing, converting, etc. (Note: on the Wii, hooks are not processed
since there is no underlying OS or command interpreter to run them.)

#### Example

For users with non-US keyboards, some keys may not work properly. This can be
worked around by switching temporarily to a US keyboard layout on startup, and
resetting the keyboard on exit. To define hooks to accomplish this:

    cat >~/.schism/startup-hook <<EOF
    #!/bin/sh
    setxkbmap us
    EOF
    cat >~/.schism/exit-hook <<EOF
    #!/bin/sh
    setxkbmap fi
    EOF
    chmod +x ~/.schism/*-hook

This is for a Finnish keyboard; replace the `fi` with the appropriate [ISO
3166](http://www.wikipedia.org/wiki/ISO%203166-1%20alpha-2) country code or
other keyboard mapping. See `/etc/X11/xkb/symbols` for a list of available
keyboard layouts.


schism/config.c
/*
 * Schism Tracker - a cross-platform Impulse Tracker clone
 * copyright (c) 2003-2005 Storlek <storlek@rigelseven.com>
 * copyright (c) 2005-2008 Mrs. Brisby <mrs.brisby@nimh.org>
 * copyright (c) 2009 Storlek & Mrs. Brisby
 * copyright (c) 2010-2012 Storlek
 * URL: http://schismtracker.org/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "headers.h"

#include "it.h"
#include "video.h" /* shouldn't need this */

#include <sys/types.h>
#include <sys/stat.h>

#include "config-parser.h"
#include "dmoz.h"
#include "osdefs.h"

/* --------------------------------------------------------------------- */
/* config settings */

char cfg_dir_modules[PATH_MAX + 1], cfg_dir_samples[PATH_MAX + 1], cfg_dir_instruments[PATH_MAX + 1],
	cfg_dir_dotschism[PATH_MAX + 1], cfg_font[NAME_MAX + 1];
char cfg_video_driver[65];
int cfg_video_fullscreen = 0;
int cfg_video_mousecursor = MOUSE_EMULATED;
int cfg_video_gl_bilinear = 1;
int cfg_video_width, cfg_video_height;

/* --------------------------------------------------------------------- */

#if defined(WIN32)
# define DOT_SCHISM "Schism Tracker"
#elif defined(MACOSX)
# define DOT_SCHISM "Library/Application Support/Schism Tracker"
#elif defined(GEKKO)
# define DOT_SCHISM "."
#else
# define DOT_SCHISM ".schism"
#endif

void cfg_init_dir(void)
{
#if defined(__amigaos4__)
	strcpy(cfg_dir_dotschism, "PROGDIR:");
#else
	char *dot_dir, *ptr;

	dot_dir = get_dot_directory();
	ptr = dmoz_path_concat(dot_dir, DOT_SCHISM);
	strncpy(cfg_dir_dotschism, ptr, PATH_MAX);
	cfg_dir_dotschism[PATH_MAX] = 0;
	free(dot_dir);
	free(ptr);

	if (!is_directory(cfg_dir_dotschism)) {
		printf("Creating directory %s\n", cfg_dir_dotschism);
		printf("Schism Tracker uses this directory to store your settings.\n");
		if (mkdir(cfg_dir_dotschism, 0777) != 0) {
			perror("Error creating directory");
			fprintf(stderr, "Everything will still work, but preferences will not be saved.\n");
		}
	}
#endif
}

/* --------------------------------------------------------------------- */

static const char palette_trans[64] = ".0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz";
static void cfg_load_palette(cfg_file_t *cfg)
{
	uint8_t colors[48];
	int n;
	char palette_text[49] = "";
	const char *ptr;

	palette_load_preset(cfg_get_number(cfg, "General", "palette", 2));

	cfg_get_string(cfg, "General", "palette_cur", palette_text, 48, "");
	for (n = 0; n < 48; n++) {
		if (palette_text[n] == '\0' || (ptr = strchr(palette_trans, palette_text[n])) == NULL)
			return;
		colors[n] = ptr - palette_trans;
	}
	memcpy(current_palette, colors, sizeof(current_palette));
}

static void cfg_save_palette(cfg_file_t *cfg)
{
	int n;
	char palette_text[49] = "";

	cfg_set_number(cfg, "General", "palette", current_palette_index);

	for (n = 0; n < 48; n++) {
		/* Changed older implementation for this, since it is not vital
		to have speed here and the compiler printed a warning */
		palette_text[n] = palette_trans[current_palette[n/3][n%3]];
	}
	palette_text[48] = '\0';
	cfg_set_string(cfg, "General", "palette_cur", palette_text);
}

/* --------------------------------------------------------------------------------------------------------- */

void cfg_load(void)
{
	char *tmp;
	const char *ptr;
	int i;
	cfg_file_t cfg;

	tmp = dmoz_path_concat(cfg_dir_dotschism, "config");
	cfg_init(&cfg, tmp);
	free(tmp);

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	cfg_get_string(&cfg, "Video", "driver", cfg_video_driver, 64, "");
	cfg_video_width = cfg_get_number(&cfg, "Video", "width", 640);
	cfg_video_height = cfg_get_number(&cfg, "Video", "height", 400);
	cfg_video_fullscreen = !!cfg_get_number(&cfg, "Video", "fullscreen", 0);
	cfg_video_mousecursor = cfg_get_number(&cfg, "Video", "mouse_cursor", MOUSE_EMULATED);
	cfg_video_mousecursor = CLAMP(cfg_video_mousecursor, 0, MOUSE_MAX_STATE);
	cfg_video_gl_bilinear =
		!!cfg_get_number(&cfg, "Video", "gl_bilinear", 1);
	ptr = cfg_get_string(&cfg, "Video", "aspect", NULL, 0, NULL);
	if (ptr && *ptr)
		put_env_var("SCHISM_VIDEO_ASPECT", ptr);

	tmp = get_home_directory();
	cfg_get_string(&cfg, "Directories", "modules", cfg_dir_modules, PATH_MAX, tmp);
	cfg_get_string(&cfg, "Directories", "samples", cfg_dir_samples, PATH_MAX, tmp);
	cfg_get_string(&cfg, "Directories", "instruments", cfg_dir_instruments, PATH_MAX, tmp);
	free(tmp);

	ptr = cfg_get_string(&cfg, "Directories", "module_pattern", NULL, 0, NULL);
	if (ptr) {
		strncpy(cfg_module_pattern, ptr, PATH_MAX);
		cfg_module_pattern[PATH_MAX] = 0;
	}

	ptr = cfg_get_string(&cfg, "Directories", "export_pattern", NULL, 0, NULL);
	if (ptr) {
		strncpy(cfg_export_pattern, ptr, PATH_MAX);
		cfg_export_pattern[PATH_MAX] = 0;
	}

	ptr = cfg_get_string(&cfg, "General", "numlock_setting", NULL, 0, NULL);
	if (!ptr)
		status.fix_numlock_setting = NUMLOCK_GUESS;
	else if (strcasecmp(ptr, "on") == 0)
		status.fix_numlock_setting = NUMLOCK_ALWAYS_ON;
	else if (strcasecmp(ptr, "off") == 0)
		status.fix_numlock_setting = NUMLOCK_ALWAYS_OFF;
	else
		status.fix_numlock_setting = NUMLOCK_HONOR;

	set_key_repeat(cfg_get_number(&cfg, "General", "key_repeat_delay", key_repeat_delay()),
		       cfg_get_number(&cfg, "General", "key_repeat_rate", key_repeat_rate()));

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	cfg_load_info(&cfg);
	cfg_load_patedit(&cfg);
	cfg_load_audio(&cfg);
	cfg_load_midi(&cfg);
	cfg_load_disko(&cfg);
	cfg_load_dmoz(&cfg);

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	if (cfg_get_number(&cfg, "General", "classic_mode", 0))
		status.flags |= CLASSIC_MODE;
	else
		status.flags &= ~CLASSIC_MODE;
	if (cfg_get_number(&cfg, "General", "make_backups", 1))
		status.flags |= MAKE_BACKUPS;
	else
		status.flags &= ~MAKE_BACKUPS;
	if (cfg_get_number(&cfg, "General", "numbered_backups", 0))
		status.flags |= NUMBERED_BACKUPS;
	else
		status.flags &= ~NUMBERED_BACKUPS;

	i = cfg_get_number(&cfg, "General", "time_display", TIME_PLAY_ELAPSED);
	/* default to play/elapsed for invalid values */
	if (i < 0 || i >= TIME_PLAYBACK)
		i = TIME_PLAY_ELAPSED;
	status.time_display = i;

	i = cfg_get_number(&cfg, "General", "vis_style", VIS_OSCILLOSCOPE);
	/* default to oscilloscope for invalid values */
	if (i < 0 || i >= VIS_SENTINEL)
		i = VIS_OSCILLOSCOPE;
	status.vis_style = i;

	kbd_sharp_flat_toggle(cfg_get_number(&cfg, "General", "accidentals_as_flats", 0) == 1);

#ifdef MACOSX
# define DEFAULT_META 1
#else
# define DEFAULT_META 0
#endif
	if (cfg_get_number(&cfg, "General", "meta_is_ctrl", DEFAULT_META))
		status.flags |= META_IS_CTRL;
	else
		status.flags &= ~META_IS_CTRL;
	if (cfg_get_number(&cfg, "General", "altgr_is_alt", 1))
		status.flags |= ALTGR_IS_ALT;
	else
		status.flags &= ~ALTGR_IS_ALT;
	if (cfg_get_number(&cfg, "Video", "lazy_redraw", 0))
		status.flags |= LAZY_REDRAW;
	else
		status.flags &= ~LAZY_REDRAW;

	if (cfg_get_number(&cfg, "General", "midi_like_tracker", 0))
		status.flags |= MIDI_LIKE_TRACKER;
	else
		status.flags &= ~MIDI_LIKE_TRACKER;

	cfg_get_string(&cfg, "General", "font", cfg_font, NAME_MAX, "font.cfg");

	cfg_load_palette(&cfg);

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
	/* Extended flags */

    if (cfg_get_number(&cfg, "General", "bypass_splash", 0))
		status.flags_extended |= BYPASS_SPLASH;
	else
		status.flags_extended &= ~BYPASS_SPLASH;

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	cfg_free(&cfg);
}

void cfg_midipage_save(void)
{
	char *ptr;
	cfg_file_t cfg;

	ptr = dmoz_path_concat(cfg_dir_dotschism, "config");
	cfg_init(&cfg, ptr);
	free(ptr);

	cfg_save_midi(&cfg);

	cfg_write(&cfg);
	cfg_free(&cfg);
}

void cfg_save(void)
{
	char *ptr;
	cfg_file_t cfg;

	ptr = dmoz_path_concat(cfg_dir_dotschism, "config");
	cfg_init(&cfg, ptr);
	free(ptr);

	// this wart is here because Storlek is retarded
	cfg_delete_key(&cfg, "Directories", "filename_pattern");

	cfg_set_string(&cfg, "Directories", "modules", cfg_dir_modules);
	cfg_set_string(&cfg, "Directories", "samples", cfg_dir_samples);
	cfg_set_string(&cfg, "Directories", "instruments", cfg_dir_instruments);
	/* No, it's not a directory, but whatever. */
	cfg_set_string(&cfg, "Directories", "module_pattern", cfg_module_pattern);
	cfg_set_string(&cfg, "Directories", "export_pattern", cfg_export_pattern);

	cfg_save_info(&cfg);
	cfg_save_patedit(&cfg);
	cfg_save_audio(&cfg);
	cfg_save_palette(&cfg);
	cfg_save_disko(&cfg);
	cfg_save_dmoz(&cfg);

	cfg_write(&cfg);
	cfg_free(&cfg);
}

void cfg_atexit_save(void)
{
	char *ptr;
	cfg_file_t cfg;

	ptr = dmoz_path_concat(cfg_dir_dotschism, "config");
	cfg_init(&cfg, ptr);
	free(ptr);

	cfg_atexit_save_audio(&cfg);

	/* TODO: move these config options to video.c, this is lame :)
	Or put everything here, which is what the note in audio_loadsave.cc
	says. Very well, I contradict myself. */
	cfg_set_string(&cfg, "Video", "driver", video_driver_name());
	cfg_set_number(&cfg, "Video", "fullscreen", !!(video_is_fullscreen()));
	cfg_set_number(&cfg, "Video", "mouse_cursor", video_mousecursor_visible());
	cfg_set_number(&cfg, "Video", "gl_bilinear", video_gl_bilinear());
	cfg_set_number(&cfg, "Video", "lazy_redraw", !!(status.flags & LAZY_REDRAW));

	cfg_set_number(&cfg, "General", "vis_style", status.vis_style);
	cfg_set_number(&cfg, "General", "time_display", status.time_display);
	cfg_set_number(&cfg, "General", "classic_mode", !!(status.flags & CLASSIC_MODE));
	cfg_set_number(&cfg, "General", "make_backups", !!(status.flags & MAKE_BACKUPS));
	cfg_set_number(&cfg, "General", "numbered_backups", !!(status.flags & NUMBERED_BACKUPS));

	cfg_set_number(&cfg, "General", "accidentals_as_flats", !!(status.flags & ACCIDENTALS_AS_FLATS));
	cfg_set_number(&cfg, "General", "meta_is_ctrl", !!(status.flags & META_IS_CTRL));
	cfg_set_number(&cfg, "General", "altgr_is_alt", !!(status.flags & ALTGR_IS_ALT));

	cfg_set_number(&cfg, "General", "midi_like_tracker", !!(status.flags & MIDI_LIKE_TRACKER));
	/* Say, whose bright idea was it to make this a string setting?
	The config file is human editable but that's mostly for developer convenience and debugging
	purposes. These sorts of things really really need to be options in the GUI so that people
	don't HAVE to touch the settings. Then we can just use an enum (and we *could* in theory
	include comments to the config by default listing what the numbers are, but that shouldn't
	be necessary in most cases. */
	switch (status.fix_numlock_setting) {
	case NUMLOCK_ALWAYS_ON:
		cfg_set_string(&cfg, "General", "numlock_setting", "on");
		break;
	case NUMLOCK_ALWAYS_OFF:
		cfg_set_string(&cfg, "General", "numlock_setting", "off");
		break;
	case NUMLOCK_HONOR:
		cfg_set_string(&cfg, "General", "numlock_setting", "system");
		break;
	case NUMLOCK_GUESS:
		/* leave empty */
		break;
	};

	/* hm... most of the time probably nothing's different, so saving the
	config file here just serves to make the backup useless. maybe add a
	'dirty' flag to the config parser that checks if any settings are
	actually *different* from those in the file? */

	cfg_write(&cfg);
	cfg_free(&cfg);
}
